aggregator/src/aggregation/blob_data.rs

```diff
@@ -186,86 +195,96 @@ impl BlobDataConfig {
+            |meta| {
+                let is_data = meta.query_selector(config.data_selector);
+                let is_hash = meta.query_selector(config.hash_selector);
+                let is_not_hash = 1.expr() - is_hash;
+                let is_boundary = meta.query_advice(config.is_boundary, Rotation::cur());
 
-            let hash_section_table = vec![
-                meta.query_selector(config.hash_selector),
-                meta.query_advice(config.chunk_idx, Rotation::cur()),
-                meta.query_advice(config.accumulator, Rotation::cur()),
-                meta.query_advice(config.digest_rlc, Rotation::cur()),
-            ];
-            [
-                1.expr(),                                               // hash section
-                meta.query_advice(config.chunk_idx, Rotation::cur()),   // chunk idx
-                meta.query_advice(config.accumulator, Rotation::cur()), // chunk len
-                meta.query_advice(config.digest_rlc, Rotation::cur()),  // digest rlc
-            ]
-            .into_iter()
-            .zip(hash_section_table)
-            .map(|(value, table)| (cond.expr() * value, table))
-            .collect()
-        });
+                // in the "metadata" or "chunk data" section, wherever is_boundary is set.
+                let cond = is_not_hash * is_boundary;
+
+                let accumulator = meta.query_advice(config.accumulator, Rotation::cur());
+                let preimage_len =
+                    is_data.expr() * accumulator + (1.expr() - is_data) * N_ROWS_METADATA.expr();
+
+                [
+                    1.expr(),                                                // q_enable
+                    1.expr(),                                                // is final
+                    meta.query_advice(config.preimage_rlc, Rotation::cur()), // input RLC
+                    preimage_len,                                            // input len
+                    meta.query_advice(config.digest_rlc, Rotation::cur()),   // output RLC
+                ]
+                .into_iter()
+                .zip_eq(keccak_table.table_exprs(meta))
+                .map(|(value, table)| (cond.expr() * value, table))
+                .collect()
+            },
+        );
 
-        // lookup for challenge_digest := keccak(preimage_challenge_digest)
-        meta.lookup_any("BlobDataConfig (z := keccak(preimage_z))", |meta| {
-            let is_hash = meta.query_selector(config.hash_selector);
-            let is_boundary = meta.query_advice(config.is_boundary, Rotation::cur());
+        // lookup chunk data digests in the "digest rlc section" of BlobDataConfig.
+        meta.lookup_any(
+            "BlobDataConfig (chunk data digests in BlobDataConfig \"hash section\")",
+            |meta| {
+                let is_data = meta.query_selector(config.data_selector);
+                let is_boundary = meta.query_advice(config.is_boundary, Rotation::cur());
```

Comment:
Do we have mechanism to prevent malicious prover from skipping valid and non-empty chunk? For example, if non-empty chunk i is skipped in "data section", then the i-th digest_rlc row in hash section does not necessarily be correct.